File: features/step_definitions/tools_steps.cpp
#include <gtest/gtest.h>
#include <cucumber-cpp/autodetect.hpp>
#include "../../src/tools.h"
#include <vector>
#include "helpers.h"

using cucumber::ScenarioScope;
using Eigen::Map;
using Eigen::VectorXd;

struct ToolsCtx {
  Tools tools;
  VectorXd result;
  std::vector<VectorXd> estimations;
  std::vector<VectorXd> ground_truth;
};


GIVEN("^estimations length is 0$") {
  ScenarioScope<ToolsCtx> context;
  context->estimations.clear();
}


GIVEN("^ground_truth length is 0$") {
  ScenarioScope<ToolsCtx> context;
  context->estimations.clear();
}


GIVEN("^estimations is$", (const std::string estimations_str)) {
  ScenarioScope<ToolsCtx> context;
  context->estimations = ParseVectorOfVectorXd(estimations_str);
}


GIVEN("^ground_truth is$", (const std::string ground_truth_str)) {
  ScenarioScope<ToolsCtx> context;
  context->ground_truth = ParseVectorOfVectorXd(ground_truth_str);
}


WHEN("^I calculate RMSE$") {
  ScenarioScope<ToolsCtx> context;
  context->result = context->tools.CalculateRMSE(context->estimations, context->ground_truth);
}


THEN("^I should get \"([^\"]*)\" as output$", (const std::string output_str)) {
  VectorXd expected_rmse = ParseVectorXd(output_str);
  ScenarioScope<ToolsCtx> context;
  EXPECT_TRUE(context->result.isApprox(expected_rmse));
}

File: features/steps/tools_steps.py
import sys

sys.path.append("./python")

from behave import *
from ukf.tools import Tools
from ukf.matrix import Matrix
from numpy.testing import assert_array_almost_equal
from helpers import *

@given(u'estimations length is 0')
def step_impl(context):
    context.estimations = []


@given(u'ground_truth length is 0')
def step_impl(context):
    context.ground_truth = []


@given(u'ground_truth is')
def step_impl(context):
    context.ground_truth = [parse_vector(line) for line in context.text.splitlines()]


@when(u'I calculate RMSE')
def step_impl(context):
    tools = Tools()
    context.result = tools.calculate_rmse(context.estimations,
                                          context.ground_truth)


@then(u'I should get "{rmse}" as output')
def step_impl(context, rmse):
    assert_array_almost_equal(context.result.value,
                              [[float(i)] for i in rmse.split()])


@given(u'estimations is')
def step_impl(context):
    context.estimations = [parse_vector(line) for line in context.text.splitlines()]